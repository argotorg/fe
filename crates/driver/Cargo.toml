[package]
name = "fe-driver"
version = "0.13.0-alpha"
authors = ["The Fe Developers <snakecharmers@ethereum.org>"]
edition = "2021"
license = "GPL-3.0-or-later"
repository = "https://github.com/ethereum/fe"

[features]
# Includes the solidity compiler for Yul -> bytecode compilation.
solc-backend = ["fe-yulc"]

[dependencies]
serde_json = "1.0"
fe-abi = {path = "../abi", version = "^0.13.0-alpha"}
fe-analyzer = {path = "../analyzer", version = "^0.13.0-alpha"}
fe-common = {path = "../common", version = "^0.13.0-alpha"}
fe-lowering = {path = "../lowering", version = "^0.13.0-alpha"}
fe-mir = {path = "../mir", version = "^0.13.0-alpha"}
fe-parser = {path = "../parser", version = "^0.13.0-alpha"}
fe-yulgen = {path = "../yulgen", version = "^0.13.0-alpha"}
fe-yulc = {path = "../yulc", version = "^0.13.0-alpha", features = ["solc-backend"], optional = true}
indexmap = "1.6.2"
vfs = "0.5.1"
smol_str = "0.1.21"
toml = "0.5.8"
