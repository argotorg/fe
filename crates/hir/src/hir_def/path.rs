use crate::{hir_def::Partial, HirDb};

use super::{kw, IdentId};

#[salsa::interned]
pub struct PathId {
    #[return_ref]
    pub segments: Vec<Partial<IdentId>>,
}

impl PathId {
    pub fn last_segment(self, db: &dyn HirDb) -> Partial<IdentId> {
        self.segments(db).last().copied().unwrap_or_default()
    }

    pub fn len(self, db: &dyn HirDb) -> usize {
        self.segments(db).len()
    }

    pub fn self_ty(db: &dyn HirDb) -> Self {
        let self_ty = Partial::Present(kw::SELF_TY);
        Self::new(db, vec![self_ty])
    }

    pub fn from_ident(db: &dyn HirDb, ident: IdentId) -> Self {
        Self::new(db, vec![Partial::Present(ident)])
    }
}
