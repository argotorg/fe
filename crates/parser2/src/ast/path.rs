use rowan::ast::{support, AstNode};

use super::{ast_node, AstChildren};
use crate::{syntax_node::SyntaxToken, SyntaxKind as SK};

ast_node! {
    /// A path.
    /// `foo::bar::baz`
    pub struct Path,
    SK::Path,
    IntoIterator<Item=PathSegment>,
}
impl Path {
    /// Returns the segments of the path.
    pub fn segments(&self) -> AstChildren<PathSegment> {
        support::children(self.syntax())
    }
}

ast_node! {
    /// A path segment.
    pub struct PathSegment,
    SK::PathSegment
}
impl PathSegment {
    /// Returns the identifier of the segment.
    pub fn ident(&self) -> Option<SyntaxToken> {
        support::token(self.syntax(), SK::Ident)
    }

    /// Returns `true` if the segment  is a `self` keyword.
    pub fn is_self(&self) -> bool {
        support::token(self.syntax(), SK::SelfKw).is_some()
    }

    /// Returns `true` if the segment is a `Self` keyword.
    pub fn is_self_ty(&self) -> bool {
        support::token(self.syntax(), SK::SelfTypeKw).is_some()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        lexer::Lexer,
        parser::{path::PathScope, Parser},
    };

    use wasm_bindgen_test::wasm_bindgen_test;

    fn parse_path(source: &str) -> Path {
        let lexer = Lexer::new(source);
        let mut parser = Parser::new(lexer);
        parser.parse(PathScope::default(), None);
        Path::cast(parser.finish().0).unwrap()
    }

    #[test]
    #[wasm_bindgen_test]
    fn path_ast() {
        let source = r#"self::Foo"#;
        let path = parse_path(source);
        let mut segments = path.segments();

        assert!(segments.next().unwrap().is_self());
        assert_eq!(segments.next().unwrap().ident().unwrap().text(), "Foo");
        assert!(segments.next().is_none());
    }

    #[test]
    #[wasm_bindgen_test]
    fn path_ast2() {
        let source = r#"Self::Dep"#;
        let path = parse_path(source);
        let mut segments = path.segments();

        assert!(segments.next().unwrap().is_self_ty());
        assert_eq!(segments.next().unwrap().ident().unwrap().text(), "Dep");
        assert!(segments.next().is_none());
    }
}
