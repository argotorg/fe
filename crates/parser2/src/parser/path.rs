use crate::SyntaxKind;

use super::{define_scope, token_stream::TokenStream, Parser};

define_scope! {
    #[doc(hidden)]
    pub PathScope,
    Path,
    Inheritance
}
impl super::Parse for PathScope {
    fn parse<S: TokenStream>(&mut self, parser: &mut Parser<S>) {
        parser.set_newline_as_trivia(false);
        parser.parse(PathSegmentScope::default(), None);
        while parser.current_kind() == Some(SyntaxKind::Colon2) {
            parser.bump_expected(SyntaxKind::Colon2);
            parser.parse(PathSegmentScope::default(), None);
        }
    }
}

define_scope! {
    PathSegmentScope,
    PathSegment,
    Inheritance
}
impl super::Parse for PathSegmentScope {
    fn parse<S: TokenStream>(&mut self, parser: &mut Parser<S>) {
        match parser.current_kind() {
            Some(kind) if is_path_segment(kind) => {
                parser.bump();
            }
            _ => parser.error_and_recover("expected path segment", None),
        }
    }
}

pub(super) fn is_path_segment(kind: SyntaxKind) -> bool {
    matches!(
        kind,
        SyntaxKind::SelfType | SyntaxKind::SelfKw | SyntaxKind::Ident
    )
}
