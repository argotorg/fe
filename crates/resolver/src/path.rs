use camino::Utf8PathBuf;
use serde::Deserialize;

use crate::remote::GitDescription;

#[derive(Deserialize, Hash, Debug, Clone, PartialEq, Eq)]
pub struct PathDescription {
    pub path: String,
}

#[derive(Clone, Debug)]
enum PathHead {
    Local(Utf8PathBuf),
    Remote(GitDescription),
}

#[derive(Clone, Debug)]
pub struct FullPathDescription {
    head: PathHead,
    sub_paths: Vec<PathDescription>,
}

impl FullPathDescription {
    pub fn new_local(local_path: &Utf8PathBuf) -> Self {
        FullPathDescription {
            head: PathHead::Local(local_path.clone()),
            sub_paths: vec![],
        }
    }

    pub fn new_remote(git_description: &GitDescription) -> Self {
        FullPathDescription {
            head: PathHead::Remote(git_description.clone()),
            sub_paths: vec![],
        }
    }

    pub fn path(&self) -> Utf8PathBuf {
        match &self.head {
            PathHead::Local(local_path) => {
                self.sub_paths
                    .iter()
                    .fold(local_path.clone(), |acc, path_description| {
                        acc.join(Utf8PathBuf::from(path_description.path.to_owned()))
                            .canonicalize_utf8()
                            .unwrap()
                    })
            }
            PathHead::Remote(git_description) => self.sub_paths.iter().fold(
                git_description.relative_path(),
                |acc, path_description| {
                    acc.join(Utf8PathBuf::from(path_description.path.to_owned()))
                        .canonicalize_utf8()
                        .unwrap()
                },
            ),
        }
    }

    pub fn push_sub_path(&self, path_description: &PathDescription) -> Self {
        let mut sub_paths = self.sub_paths.clone();
        sub_paths.push(path_description.clone());
        Self {
            head: self.head.clone(),
            sub_paths,
        }
    }
}
