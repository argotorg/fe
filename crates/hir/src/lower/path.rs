use fe_parser2::ast;

use crate::{
    hir_def::{IdentId, MaybeInvalid, PathId, PathSegment},
    HirDb,
};

impl PathId {
    pub(crate) fn from_ast(db: &dyn HirDb, ast: ast::Path) -> Self {
        let mut segments = Vec::new();
        for seg in ast.into_iter() {
            let segment = if seg.is_self() {
                MaybeInvalid::Valid(PathSegment::Self_)
            } else if seg.is_self_ty() {
                MaybeInvalid::Valid(PathSegment::SelfTy)
            } else if let Some(ident) = seg.ident() {
                MaybeInvalid::Valid(PathSegment::Ident(IdentId::new(
                    db,
                    ident.text().to_string(),
                )))
            } else {
                MaybeInvalid::invalid()
            };
            segments.push(segment);
        }

        Self::new(db, segments)
    }

    pub(crate) fn maybe_from_ast(db: &dyn HirDb, ast: Option<ast::Path>) -> MaybeInvalid<Self> {
        ast.map(|ast| Self::from_ast(db, ast)).into()
    }
}
