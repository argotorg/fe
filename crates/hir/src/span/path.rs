use parser::ast;

use super::define_lazy_span_item;

define_lazy_span_item!(
    LazyPathSpan,
    ast::Path,
    @idx {
        (segment, LazyPathSegmentSpan),
    }
);

define_lazy_span_item!(LazyPathSegmentSpan);
