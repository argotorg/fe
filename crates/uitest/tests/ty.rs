use std::path::Path;

use dir_test::{dir_test, Fixture};
use driver::DriverDataBase;
use fe_compiler_test_utils::snap_test;

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/fixtures/ty/def",
    glob: "*.fe"
)]
fn run_ty_def(fixture: Fixture<&str>) {
    let mut driver = DriverDataBase::default();
    let path = Path::new(fixture.path());
    let top_mod = driver.top_mod_from_file(path, fixture.content());
    driver.run_on_top_mod(top_mod);
    let diags = driver.format_diags();
    snap_test!(diags, fixture.path());
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/fixtures/ty/const_ty",
    glob: "*.fe"
)]
fn run_const_ty(fixture: Fixture<&str>) {
    let mut driver = DriverDataBase::default();
    let path = Path::new(fixture.path());
    let top_mod = driver.top_mod_from_file(path, fixture.content());
    driver.run_on_top_mod(top_mod);
    let diags = driver.format_diags();
    snap_test!(diags, fixture.path());
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/fixtures/ty/trait_bound",
    glob: "*.fe"
)]
fn run_trait_bound(fixture: Fixture<&str>) {
    let mut driver = DriverDataBase::default();
    let path = Path::new(fixture.path());
    let top_mod = driver.top_mod_from_file(path, fixture.content());
    driver.run_on_top_mod(top_mod);
    let diags = driver.format_diags();
    snap_test!(diags, fixture.path());
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/fixtures/ty/trait_impl",
    glob: "*.fe"
)]
fn run_trait_impl(fixture: Fixture<&str>) {
    let mut driver = DriverDataBase::default();
    let path = Path::new(fixture.path());
    let top_mod = driver.top_mod_from_file(path, fixture.content());
    driver.run_on_top_mod(top_mod);
    let diags = driver.format_diags();
    snap_test!(diags, fixture.path());
}

#[cfg(target_family = "wasm")]
mod wasm {
    use wasm_bindgen_test::wasm_bindgen_test;

    use super::*;

    mod def {
        use super::*;

        // TODO: we opt out the tests for type/alias/trait-infinite recursion checks. See https://github.com/ethereum/fe/issues/939 for more details.
        #[dir_test(
        dir: "$CARGO_MANIFEST_DIR/fixtures/ty/def",
        glob: "*[!_cycle].fe",
        postfix: "wasm"
        )]
        #[dir_test_attr(
            #[wasm_bindgen_test]
        )]
        fn run_ty_def(fixture: Fixture<&str>) {
            let mut driver = DriverDataBase::default();
            let path = Path::new(fixture.path());
            let top_mod = driver.top_mod_from_file(path, fixture.content());
            driver.run_on_top_mod(top_mod);
        }
    }

    mod const_ty {
        use super::*;

        #[dir_test(
        dir: "$CARGO_MANIFEST_DIR/fixtures/ty/const_ty",
        glob: "*.fe",
        postfix: "wasm"
        )]
        #[dir_test_attr(
            #[wasm_bindgen_test]
        )]
        fn run_const_ty(fixture: Fixture<&str>) {
            let mut driver = DriverDataBase::default();
            let path = Path::new(fixture.path());
            let top_mod = driver.top_mod_from_file(path, fixture.content());
            driver.run_on_top_mod(top_mod);
        }
    }

    mod trait_bound {
        use super::*;

        #[dir_test(
        dir: "$CARGO_MANIFEST_DIR/fixtures/ty/trait_bound",
        glob: "*.fe",
        postfix: "wasm"
        )]
        #[dir_test_attr(
            #[wasm_bindgen_test]
        )]
        fn run_trait_bound(fixture: Fixture<&str>) {
            let mut driver = DriverDataBase::default();
            let path = Path::new(fixture.path());
            let top_mod = driver.top_mod_from_file(path, fixture.content());
            driver.run_on_top_mod(top_mod);
        }
    }

    mod trait_impl {
        use super::*;

        #[dir_test(
        dir: "$CARGO_MANIFEST_DIR/fixtures/ty/trait_impl",
        glob: "*.fe",
        postfix: "wasm"
        )]
        #[dir_test_attr(
            #[wasm_bindgen_test]
        )]
        fn run_trait_impl(fixture: Fixture<&str>) {
            let mut driver = DriverDataBase::default();
            let path = Path::new(fixture.path());
            let top_mod = driver.top_mod_from_file(path, fixture.content());
            driver.run_on_top_mod(top_mod);
        }
    }
}
