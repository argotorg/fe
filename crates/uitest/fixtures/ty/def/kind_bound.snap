---
source: crates/uitest/tests/ty.rs
expression: diags
input_file: crates/uitest/fixtures/ty/kind_bound.fe
---
error[2-0002]: `wrapper3` is not found
   ┌─ kind_bound.fe:34:15
   │
34 │     bar_err3: wrapper3<Wrapper1, Wrapper1>,
   │               ^^^^^^^^ `wrapper3` is not found

error[3-0001]: invalid type argument kind
   ┌─ kind_bound.fe:25:14
   │
25 │     foo_err: Wrapper2<i32, i32>,
   │              ^^^^^^^^^^^^^^^^^^ expected `(* -> *)` kind, but `i32` has `*` kind

error[3-0001]: invalid type argument kind
   ┌─ kind_bound.fe:32:15
   │
32 │     bar_err1: Wrapper3<i32, u32>,
   │               ^^^^^^^^^^^^^^^^^^ expected `((* -> *) -> (* -> *))` kind, but `i32` has `*` kind

error[3-0001]: invalid type argument kind
   ┌─ kind_bound.fe:33:15
   │
33 │     bar_err2: Wrapper3<Wrapper2, u32>,
   │               ^^^^^^^^^^^^^^^^^^^^^^^ expected `(* -> *)` kind, but `u32` has `*` kind

error[3-0005]: duplicate type bound is not allowed.
   ┌─ kind_bound.fe:38:10
   │
38 │ where T: (* -> *) -> *
   │          ^^^^^^^^^^^^^ `T` is already declared with `(* -> *)` kind, but found `((* -> *) -> *)` kind here

error[6-0005]: trait bound for concrete type is not allowed
   ┌─ kind_bound.fe:44:7
   │
44 │ where Self: *
   │       ^^^^ `InvalidBound2` is a concrete type

error[6-0005]: trait bound for concrete type is not allowed
   ┌─ kind_bound.fe:50:7
   │
50 │ where Wrapper1<i32>: *
   │       ^^^^^^^^^^^^^ `Wrapper1<i32>` is a concrete type


